pkg/acl/cache.go

```diff
@@ -0,0 +1,140 @@
+/*
+ * JuiceFS, Copyright 2024 Juicedata, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package acl
+
+import (
+	"sync"
+)
+
+const None = 0
+
+// Cache all rules
+// - cache all rules when meta init.
+// - on getfacl failure, read and cache rule from meta.
+// - on setfacl success, read and cache all missed rules from meta. (considered as a low-frequency operation)
+// - concurrent mounts may result in duplicate rules.
+type Cache interface {
+	Put(id uint32, r *Rule)
+	Get(id uint32) *Rule
+	GetId(r *Rule) uint32
+	Size() int
+	GetMissIds(maxId uint32) []uint32
+}
+
+func NewCache() Cache {
+	return &cache{
+		lock:     sync.RWMutex{},
+		maxId:    None,
+		id2Rule:  make(map[uint32]*Rule),
+		cksum2Id: make(map[uint32][]uint32),
+	}
+}
+
+type cache struct {
+	lock     sync.RWMutex
+	maxId    uint32
+	id2Rule  map[uint32]*Rule
+	cksum2Id map[uint32][]uint32
+}
+
+func (c *cache) GetMissIds(maxId uint32) []uint32 {
```

Comment:
Assume all the IDs are continuous, should we scan the database with prefix or table, rather than generating missing ids?

Reply:
case: In a multi-client concurrent situation, Client A initializes with IDs 1-10. Client B sets ACLs for IDs 11-20. Next, Client A needs to fetch the missing ACLs for IDs 11-20.

Reply:
Client can fetch missed one by one when need

Reply:
During ACL insertion into metadata, the client can only rely on local cache to determine the existence of ACL values, which may result in duplicate ACL values. Therefore, I currently batch read to the local cache during ACL insertion if the ACL is missing, reducing the likelihood of duplicate ACLs in the future. Since insertion should be a low-frequency operation, this overhead is acceptable